<a href="https://colab.research.google.com/github/Arisarayangyoo/bcs_dpdm24/blob/main/FinalExam_653020606_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import data

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
train_data = pd.read_csv('/content/drive/MyDrive/Data/train_set.csv')
train_data

id number    Xn  Hb(Gm%)  Hct(%)  MCV(fl)  A2 (2-3%)  HbF (2%)  \
0       67d169  25.0     11.1    34.0     78.0        3.2       0.0   
1       67l168  32.0     13.9    42.0     86.0        3.7       0.3   
2       67i267  43.0     10.1    31.0     78.0        2.6       0.3   
3       67e248  33.0     12.9    38.0     76.0        3.0       0.2   
4       67h277  23.0     12.9    40.0     78.0        3.1       0.0   
...        ...   ...      ...     ...      ...        ...       ...   
4317    66z431  63.0     10.1    31.0     83.0        2.8       1.8   
4318    66z436  19.0     11.1    34.0     79.0        2.6       0.6   
4319    66z439  80.0     10.9    33.9     79.0        3.0       0.2   
4320    66z445  27.0     13.2    41.4     87.0        2.9       0.3   
4321     66z66  29.0      7.2    22.9     78.0        2.4       0.0   

      HbA (95-97%)       RBC        MCH       MCHC  MCV-RBC-3*HGB  \
0             55.9  4.358974  25.464706  32.647059      40.341026   
1             54.8  5.197674  28.859060  33.557047      35.802326   
2             56.2  3.974359  25.412903  32.580645      43.725641   
3             61.2  5.000000  25.800000  33.947368      32.300000   
4             55.6  5.128205  25.155000  32.250000      34.171795   
...            ...       ...        ...        ...            ...   
4317          84.2  3.734940  27.041935  32.580645      48.965060   
4318          78.0  4.303797  25.791176  32.647059      41.396203   
4319          87.2  4.291139  25.401180  32.153392      42.008861   
4320          87.6  4.758621  27.739130  31.884058      42.641379   
4321          97.6  3.100602  24.524017  31.441048      53.464103   

      MCV-RBC-5*HGB  |80-MCV|  |27-MCH|  |80-MCV|*|27-MCH|         Thal_Group  
0         18.141026       2.0  1.535294           3.070588  Abnormal_variants  
1          5.802326       6.0  1.859060          11.154362  Abnormal_variants  
2         23.525641       2.0  1.587097           3.174194  Abnormal_variants  
3          6.500000       4.0  1.200000           4.800000  Abnormal_variants  
4          8.371795       2.0  1.845000           3.690000  Abnormal_variants  
...             ...       ...       ...                ...                ...  
4317      28.765060       3.0  0.041935           0.125806             Normal  
4318      19.196203       1.0  1.208824           1.208824             Normal  
4319      20.208861       1.0  1.598820           1.598820             Normal  
4320      16.241379       7.0  0.739130           5.173913             Normal  
4321      35.679858       2.0  2.475983           4.951965             Normal  

[4322 rows x 17 columns]

In [ ]:
X_test = pd.read_csv('/content/drive/MyDrive/Data/X_test.csv')
X_test

id number    Xn  Hb(Gm%)  Hct(%)  MCV(fl)  A2 (2-3%)  HbF (2%)  \
0       66z133  38.0     12.4   37.00     78.0        2.7       1.3   
1       66z297  28.0     12.8   39.40     82.0        2.7       0.0   
2       66u491  78.0     10.9   33.90     79.0        3.1       0.0   
3       66x072  65.0     10.9   33.90     79.0        3.0       0.2   
4       67m031  17.0     11.5   36.00     78.0        2.8       0.8   
...        ...   ...      ...     ...      ...        ...       ...   
1992    67j317   NaN      8.4   26.00     59.0        2.6       0.7   
1993    66a385   NaN     10.1   31.00     78.0        3.3       0.0   
1994    67g355  53.0     10.4   34.00     77.0        3.0       0.2   
1995    66s121  54.0     10.1   31.00     78.0        2.4       0.3   
1996    67h101  64.0     11.9   37.85     65.5       21.7       1.1   

      HbA (95-97%)     RBC      MCH     MCHC  MCV-RBC-3*HGB  MCV-RBC-5*HGB  \
0             75.4  4.7436  26.1405  33.5135        36.0564        11.2564   
1             97.3  4.8049  26.6396  32.4873        38.7951        13.1951   
2             96.9  4.2911  25.4012  32.1534        42.0089        20.2089   
3             87.9  4.2911  25.4012  32.1534        42.0089        20.2089   
4             79.6  4.6154  24.9167  31.9444        38.8846        15.8846   
...            ...     ...      ...      ...            ...            ...   
1992           0.3  4.4068  19.0615  32.3077        29.3932        12.5932   
1993          55.0  3.9744  25.4129  32.5806        43.7256        23.5256   
1994          49.6  4.4156  23.5529  30.5882        41.3844        20.5844   
1995          42.0  3.9744  25.4129  32.5806        43.7256        23.5256   
1996          66.1  5.7786  20.5931  31.4399        24.0214         0.2214   

      |80-MCV|  |27-MCH|  |80-MCV|*|27-MCH|  
0          2.0    0.8595             1.7189  
1          2.0    0.3604             0.7208  
2          1.0    1.5988             1.5988  
3          1.0    1.5988             1.5988  
4          2.0    2.0833             4.1667  
...        ...       ...                ...  
1992      21.0    7.9385           166.7077  
1993       2.0    1.5871             3.1742  
1994       3.0    3.4471            10.3412  
1995       2.0    1.5871             3.1742  
1996      14.5    6.4069            92.8996  

[1997 rows x 16 columns]

In [ ]:
y_pred = pd.read_csv('/content/drive/MyDrive/Data/y_pred_baseline.csv')
y_pred.head()

Thal_Group
0     Normal
1     Normal
2     Normal
3     Normal
4     Normal

#Preprocessing

## train_data

In [ ]:
train_data.isnull().sum()

id number              0
Xn                   435
Hb(Gm%)                7
Hct(%)                 7
MCV(fl)                7
A2 (2-3%)            133
HbF (2%)             116
HbA (95-97%)         116
RBC                    7
MCH                    7
MCHC                   7
MCV-RBC-3*HGB          7
MCV-RBC-5*HGB          7
|80-MCV|               7
|27-MCH|               7
|80-MCV|*|27-MCH|      7
Thal_Group             0
dtype: int64

In [ ]:
# prompt: จัดกลุ่มตาม Thal_Group แล้ว หาค่าเฉลี่ยของ Xn	Hb(Gm%)	Hct(%)	MCV(fl)	A2 (2-3%)	HbF (2%)	HbA (95-97%)	RBC	MCH	MCHC	MCV-RBC-3*HGB	MCV-RBC-5*HGB	|80-MCV|	|27-MCH|	|80-MCV|*|27-MCH| เติมเเทนค่าว่าง

# Group data by 'Thal_Group' and calculate the mean of specified columns
grouped_data = train_data.groupby('Thal_Group').agg({
    'Xn': 'mean',
    'Hb(Gm%)': 'mean',
    'Hct(%)': 'mean',
    'MCV(fl)': 'mean',
    'A2 (2-3%)': 'mean',
    'HbF (2%)': 'mean',
    'HbA (95-97%)': 'mean',
    'RBC': 'mean',
    'MCH': 'mean',
    'MCHC': 'mean',
    'MCV-RBC-3*HGB': 'mean',
    'MCV-RBC-5*HGB': 'mean',
    '|80-MCV|': 'mean',
    '|27-MCH|': 'mean',
    '|80-MCV|*|27-MCH|': 'mean'
}).reset_index()

# Fill missing values with the mean for each Thal_Group
for col in ['Xn', 'Hb(Gm%)', 'Hct(%)', 'MCV(fl)', 'A2 (2-3%)', 'HbF (2%)', 'HbA (95-97%)', 'RBC', 'MCH', 'MCHC', 'MCV-RBC-3*HGB', 'MCV-RBC-5*HGB', '|80-MCV|', '|27-MCH|', '|80-MCV|*|27-MCH|']:
    train_data[col] = train_data.groupby('Thal_Group')[col].transform(lambda x: x.fillna(x.mean()))
grouped_data


Thal_Group         Xn    Hb(Gm%)     Hct(%)    MCV(fl)  A2 (2-3%)  \
0   Abnormal_variants  40.529412  11.105556  33.650000  79.625000   3.227778   
1  Alpha_thal_related  41.330254   9.530366  30.960775  63.160471   6.374747   
2   Beta_thal_related  46.324937   9.790970  31.048630  63.642123   5.626389   
3   Complex_genotypes  32.666667   9.933333  31.666667  67.000000  21.166667   
4         HbE_related  35.488960  11.333333  34.428353  72.130499  34.582174   
5              Normal  41.553550  10.402556  32.443184  75.824814   2.610992   

   HbF (2%)  HbA (95-97%)       RBC        MCH       MCHC  MCV-RBC-3*HGB  \
0  0.188889     56.566667  4.240138  26.412996  33.004988      42.204306   
1  2.888781     83.691759  4.957425  19.405597  30.799821      29.253615   
2  3.213076     84.590323  4.900895  20.054667  31.505074      29.335586   
3  4.666667     70.433333  4.712341  20.945055  31.245421      32.487659   
4  1.440010     57.062782  4.821328  23.737376  32.855755      33.086737   
5  0.449891     90.748540  4.312072  24.290406  31.922800      40.048053   

   MCV-RBC-5*HGB   |80-MCV|  |27-MCH|  |80-MCV|*|27-MCH|  
0      19.870973   3.652778  1.750514           6.349985  
1      10.083931  16.859948  7.595370         137.702169  
2       9.754587  16.444635  6.957507         118.613234  
3      12.620993  13.000000  6.054945          80.061538  
4      10.338851   8.569892  3.408531          53.034388  
5      19.171919   5.849308  3.100567          40.722298

In [ ]:
train_data.isnull().sum()

id number            0
Xn                   0
Hb(Gm%)              0
Hct(%)               0
MCV(fl)              0
A2 (2-3%)            0
HbF (2%)             0
HbA (95-97%)         0
RBC                  0
MCH                  0
MCHC                 0
MCV-RBC-3*HGB        0
MCV-RBC-5*HGB        0
|80-MCV|             0
|27-MCH|             0
|80-MCV|*|27-MCH|    0
Thal_Group           0
dtype: int64

In [ ]:
train_data

id number    Xn  Hb(Gm%)  Hct(%)  MCV(fl)  A2 (2-3%)  HbF (2%)  \
0       67d169  25.0     11.1    34.0     78.0        3.2       0.0   
1       67l168  32.0     13.9    42.0     86.0        3.7       0.3   
2       67i267  43.0     10.1    31.0     78.0        2.6       0.3   
3       67e248  33.0     12.9    38.0     76.0        3.0       0.2   
4       67h277  23.0     12.9    40.0     78.0        3.1       0.0   
...        ...   ...      ...     ...      ...        ...       ...   
4317    66z431  63.0     10.1    31.0     83.0        2.8       1.8   
4318    66z436  19.0     11.1    34.0     79.0        2.6       0.6   
4319    66z439  80.0     10.9    33.9     79.0        3.0       0.2   
4320    66z445  27.0     13.2    41.4     87.0        2.9       0.3   
4321     66z66  29.0      7.2    22.9     78.0        2.4       0.0   

      HbA (95-97%)       RBC        MCH       MCHC  MCV-RBC-3*HGB  \
0             55.9  4.358974  25.464706  32.647059      40.341026   
1             54.8  5.197674  28.859060  33.557047      35.802326   
2             56.2  3.974359  25.412903  32.580645      43.725641   
3             61.2  5.000000  25.800000  33.947368      32.300000   
4             55.6  5.128205  25.155000  32.250000      34.171795   
...            ...       ...        ...        ...            ...   
4317          84.2  3.734940  27.041935  32.580645      48.965060   
4318          78.0  4.303797  25.791176  32.647059      41.396203   
4319          87.2  4.291139  25.401180  32.153392      42.008861   
4320          87.6  4.758621  27.739130  31.884058      42.641379   
4321          97.6  3.100602  24.524017  31.441048      53.464103   

      MCV-RBC-5*HGB  |80-MCV|  |27-MCH|  |80-MCV|*|27-MCH|         Thal_Group  
0         18.141026       2.0  1.535294           3.070588  Abnormal_variants  
1          5.802326       6.0  1.859060          11.154362  Abnormal_variants  
2         23.525641       2.0  1.587097           3.174194  Abnormal_variants  
3          6.500000       4.0  1.200000           4.800000  Abnormal_variants  
4          8.371795       2.0  1.845000           3.690000  Abnormal_variants  
...             ...       ...       ...                ...                ...  
4317      28.765060       3.0  0.041935           0.125806             Normal  
4318      19.196203       1.0  1.208824           1.208824             Normal  
4319      20.208861       1.0  1.598820           1.598820             Normal  
4320      16.241379       7.0  0.739130           5.173913             Normal  
4321      35.679858       2.0  2.475983           4.951965             Normal  

[4322 rows x 17 columns]

## X_test

In [ ]:
X_test.isnull().sum()

id number              0
Xn                   187
Hb(Gm%)                2
Hct(%)                 2
MCV(fl)                1
A2 (2-3%)             62
HbF (2%)              59
HbA (95-97%)          59
RBC                    2
MCH                    2
MCHC                   2
MCV-RBC-3*HGB          2
MCV-RBC-5*HGB          2
|80-MCV|               1
|27-MCH|               2
|80-MCV|*|27-MCH|      2
dtype: int64

In [ ]:
# prompt: เติมค่า missing ด้วยค่าเฉลี่ยจากข้อมูลเเต่ละคอลัม ไม่ต้องgrouped

# Fill missing values with the mean for each column in X_test
for col in X_test.columns:
    if X_test[col].isnull().any():
        mean_val = X_test[col].mean()
        X_test[col].fillna(mean_val, inplace=True)


<ipython-input-476-0eb74e4e08a4>:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  X_test[col].fillna(mean_val, inplace=True)


In [ ]:
X_test.isnull().sum()

id number            0
Xn                   0
Hb(Gm%)              0
Hct(%)               0
MCV(fl)              0
A2 (2-3%)            0
HbF (2%)             0
HbA (95-97%)         0
RBC                  0
MCH                  0
MCHC                 0
MCV-RBC-3*HGB        0
MCV-RBC-5*HGB        0
|80-MCV|             0
|27-MCH|             0
|80-MCV|*|27-MCH|    0
dtype: int64

In [ ]:
X_test

id number       Xn  Hb(Gm%)  Hct(%)  MCV(fl)  A2 (2-3%)  HbF (2%)  \
0       66z133  38.0000     12.4   37.00     78.0        2.7       1.3   
1       66z297  28.0000     12.8   39.40     82.0        2.7       0.0   
2       66u491  78.0000     10.9   33.90     79.0        3.1       0.0   
3       66x072  65.0000     10.9   33.90     79.0        3.0       0.2   
4       67m031  17.0000     11.5   36.00     78.0        2.8       0.8   
...        ...      ...      ...     ...      ...        ...       ...   
1992    67j317  40.6751      8.4   26.00     59.0        2.6       0.7   
1993    66a385  40.6751     10.1   31.00     78.0        3.3       0.0   
1994    67g355  53.0000     10.4   34.00     77.0        3.0       0.2   
1995    66s121  54.0000     10.1   31.00     78.0        2.4       0.3   
1996    67h101  64.0000     11.9   37.85     65.5       21.7       1.1   

      HbA (95-97%)     RBC      MCH     MCHC  MCV-RBC-3*HGB  MCV-RBC-5*HGB  \
0             75.4  4.7436  26.1405  33.5135        36.0564        11.2564   
1             97.3  4.8049  26.6396  32.4873        38.7951        13.1951   
2             96.9  4.2911  25.4012  32.1534        42.0089        20.2089   
3             87.9  4.2911  25.4012  32.1534        42.0089        20.2089   
4             79.6  4.6154  24.9167  31.9444        38.8846        15.8846   
...            ...     ...      ...      ...            ...            ...   
1992           0.3  4.4068  19.0615  32.3077        29.3932        12.5932   
1993          55.0  3.9744  25.4129  32.5806        43.7256        23.5256   
1994          49.6  4.4156  23.5529  30.5882        41.3844        20.5844   
1995          42.0  3.9744  25.4129  32.5806        43.7256        23.5256   
1996          66.1  5.7786  20.5931  31.4399        24.0214         0.2214   

      |80-MCV|  |27-MCH|  |80-MCV|*|27-MCH|  
0          2.0    0.8595             1.7189  
1          2.0    0.3604             0.7208  
2          1.0    1.5988             1.5988  
3          1.0    1.5988             1.5988  
4          2.0    2.0833             4.1667  
...        ...       ...                ...  
1992      21.0    7.9385           166.7077  
1993       2.0    1.5871             3.1742  
1994       3.0    3.4471            10.3412  
1995       2.0    1.5871             3.1742  
1996      14.5    6.4069            92.8996  

[1997 rows x 16 columns]

#Feature Engineering

In [ ]:
!pip install tabulate
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import cross_validate
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import (
    recall_score,
    precision_score,
    f1_score,
    classification_report
)
from imblearn.pipeline import Pipeline as imbpipeline
from imblearn.over_sampling import SMOTE
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from xgboost import XGBClassifier

## ตัด columns id number ออกเพราะไม่ไช่ Feature ที่ใช้ทำนาย

In [ ]:
train_data = train_data.drop(['id number'], axis=1)

In [ ]:
from sklearn.model_selection import train_test_split
X = train_data.drop('Thal_Group', axis=1)
y = train_data['Thal_Group']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import cross_validate, StratifiedKFold
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, make_scorer
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
# กำหนด scoring metrics สำหรับการประเมิน
scoring = {
    'accuracy': 'accuracy',
    'precision_macro': 'precision_macro',
    'recall_macro': 'recall_macro',
    'f1_macro': 'f1_macro'
}

In [ ]:
# กำหนดโมเดลที่ต้องการประเมิน
models = {
    'RandomForest': RandomForestClassifier(n_estimators=100, random_state=42),
    'LogisticRegression': LogisticRegression(max_iter=1000, random_state=42),
    'SVM': SVC(random_state=42),
    'DecisionTree': DecisionTreeClassifier(random_state=42),
    'KNN': KNeighborsClassifier()
}

In [ ]:
# สร้าง DataFrame เปล่าสำหรับเก็บผลลัพธ์
results = pd.DataFrame(columns=['Model', 'Accuracy', 'Precision', 'Recall', 'F1-score'])

In [ ]:
# ทำ Cross-Validation และเก็บผลลัพธ์
# กำหนดใช้ StratifiedKFold เพื่อรักษาสัดส่วนของแต่ละคลาสในการแบ่งข้อมูล
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

In [ ]:
# ปรับขนาดข้อมูล
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [ ]:
for name, model in models.items():
    print(f"ประเมินโมเดล {name}...")

    # ทำ cross-validation
    cv_results = cross_validate(
        model, X_scaled, y,
        cv=cv,
        scoring=scoring,
        return_train_score=False
    )

    # คำนวณค่าเฉลี่ยของผลลัพธ์จาก cross-validation
    accuracy = np.mean(cv_results['test_accuracy'])
    precision = np.mean(cv_results['test_precision_macro'])
    recall = np.mean(cv_results['test_recall_macro'])
    f1 = np.mean(cv_results['test_f1_macro'])

    # เพิ่มผลลัพธ์ลงใน DataFrame
    results = pd.concat([results, pd.DataFrame({
        'Model': [name],
        'Accuracy': [accuracy],
        'Precision': [precision],
        'Recall': [recall],
        'F1-score': [f1]
    })], ignore_index=True)


In [ ]:
print("\nผลการประเมินโมเดลด้วย 5-Fold Cross-Validation:")
pd.set_option('display.precision', 4)
print(results)


ผลการประเมินโมเดลด้วย 5-Fold Cross-Validation:
                Model  Accuracy  Precision  Recall  F1-score
0        RandomForest    0.9278     0.8405  0.7823    0.8017
1  LogisticRegression    0.7955     0.6682  0.5694    0.5857
2                 SVM    0.8785     0.6671  0.6293    0.6405
3        DecisionTree    0.9146     0.7613  0.7827    0.7699
4                 KNN    0.8635     0.7201  0.6355    0.6543


In [ ]:
# เลือกโมเดลที่ดีที่สุด (โดยพิจารณาจาก F1-score)
best_model_idx = results['F1-score'].idxmax()
best_model_name = results.loc[best_model_idx, 'Model']
print(f"\nโมเดลที่ดีที่สุดคือ: {best_model_name}")


โมเดลที่ดีที่สุดคือ: RandomForest


### โมเดลที่ดีที่สุด คือ RandomForest มีค่า F1-score สูงที่สุด 0.8017

# ทำนาย Thal_Group จาก ข้อมูลใน X_test

In [ ]:
X_test

id number       Xn  Hb(Gm%)  Hct(%)  MCV(fl)  A2 (2-3%)  HbF (2%)  \
0       66z133  38.0000     12.4   37.00     78.0        2.7       1.3   
1       66z297  28.0000     12.8   39.40     82.0        2.7       0.0   
2       66u491  78.0000     10.9   33.90     79.0        3.1       0.0   
3       66x072  65.0000     10.9   33.90     79.0        3.0       0.2   
4       67m031  17.0000     11.5   36.00     78.0        2.8       0.8   
...        ...      ...      ...     ...      ...        ...       ...   
1992    67j317  40.6751      8.4   26.00     59.0        2.6       0.7   
1993    66a385  40.6751     10.1   31.00     78.0        3.3       0.0   
1994    67g355  53.0000     10.4   34.00     77.0        3.0       0.2   
1995    66s121  54.0000     10.1   31.00     78.0        2.4       0.3   
1996    67h101  64.0000     11.9   37.85     65.5       21.7       1.1   

      HbA (95-97%)     RBC      MCH     MCHC  MCV-RBC-3*HGB  MCV-RBC-5*HGB  \
0             75.4  4.7436  26.1405  33.5135        36.0564        11.2564   
1             97.3  4.8049  26.6396  32.4873        38.7951        13.1951   
2             96.9  4.2911  25.4012  32.1534        42.0089        20.2089   
3             87.9  4.2911  25.4012  32.1534        42.0089        20.2089   
4             79.6  4.6154  24.9167  31.9444        38.8846        15.8846   
...            ...     ...      ...      ...            ...            ...   
1992           0.3  4.4068  19.0615  32.3077        29.3932        12.5932   
1993          55.0  3.9744  25.4129  32.5806        43.7256        23.5256   
1994          49.6  4.4156  23.5529  30.5882        41.3844        20.5844   
1995          42.0  3.9744  25.4129  32.5806        43.7256        23.5256   
1996          66.1  5.7786  20.5931  31.4399        24.0214         0.2214   

      |80-MCV|  |27-MCH|  |80-MCV|*|27-MCH|  
0          2.0    0.8595             1.7189  
1          2.0    0.3604             0.7208  
2          1.0    1.5988             1.5988  
3          1.0    1.5988             1.5988  
4          2.0    2.0833             4.1667  
...        ...       ...                ...  
1992      21.0    7.9385           166.7077  
1993       2.0    1.5871             3.1742  
1994       3.0    3.4471            10.3412  
1995       2.0    1.5871             3.1742  
1996      14.5    6.4069            92.8996  

[1997 rows x 16 columns]

In [ ]:
# เทรนโมเดลที่ดีที่สุดด้วยข้อมูลทั้งหมดและทำนายข้อมูล X_test
best_model = models[best_model_name]
best_model.fit(X_scaled, y)

X_test = X_test.drop(['id number'], axis=1)

# ปรับขนาดข้อมูล X_test และทำนาย
X_test_scaled = scaler.transform(X_test)
y_test_pred = best_model.predict(X_test_scaled)

In [ ]:
print(classification_report(y_test, y_pred))

                    precision    recall  f1-score   support

 Abnormal_variants       1.00      0.67      0.80         3
Alpha_thal_related       0.89      0.88      0.89       201
 Beta_thal_related       0.98      0.97      0.97        96
       HbE_related       1.00      0.99      0.99       230
            Normal       0.93      0.94      0.93       335

          accuracy                           0.94       865
         macro avg       0.96      0.89      0.92       865
      weighted avg       0.94      0.94      0.94       865



In [ ]:
predictions_df = pd.DataFrame({'Thal_Group_Predicted': y_test_pred})
predictions_df.to_csv('thal_group_predictions.csv', index=False)

print(f"\nการทำนายด้วยโมเดล {best_model_name} สำเร็จแล้ว")
print("ผลการทำนายถูกบันทึกลงในไฟล์ 'thal_group_predictions.csv'")


การทำนายด้วยโมเดล RandomForest สำเร็จแล้ว
ผลการทำนายถูกบันทึกลงในไฟล์ 'thal_group_predictions.csv'
